From patchwork Thu Apr  9 20:50:33 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 11482483
Return-Path: <SRS0=jZn3=5Z=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 0E11815AB
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 21:30:53 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id E1E452083E
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 21:30:52 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=oracle.com header.i=@oracle.com
 header.b="P915doV1"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726973AbgDIVau (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 9 Apr 2020 17:30:50 -0400
Received: from aserp2120.oracle.com ([141.146.126.78]:36656 "EHLO
        aserp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726867AbgDIVau (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 9 Apr 2020 17:30:50 -0400
Received: from pps.filterd (aserp2120.oracle.com [127.0.0.1])
        by aserp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 039LNqQZ133444;
        Thu, 9 Apr 2020 21:30:48 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references : mime-version :
 content-transfer-encoding; s=corp-2020-01-29;
 bh=MKoaiJE892ogQi9MJkgkxyHIsK17J6JTkHiivbAxBmY=;
 b=P915doV1ZHqJfsq1FuIEpBmSZts5zY5W096kfiGt+NSG6UX5ExpMPmqgcR4awCh1YkfY
 0srVFQGBeL85IY3llLJ5k3/0hX56TxG6gdYDLUTVWa1WJvOSX6slTtHuzLm2C4X7pI4Z
 KQ0rtT0wM9F/0YY4fEpGZ1RR6MKpMmZJb/t/3vmi58suDJlWuWJuynnqX+xxWGxmZDwU
 vl5+wOnSwXEnmgfoJZ6/1X8I7Qus8cMfS4UW+RtNMsrID9WWUcIROB9ScrTMVZWPO8so
 V8dFTxc2wIl9TjvbCiDRhnell96KGxMvS18cP7CfWD4OS4WYIlxrhOWYx11PFdFPQKeY 7A==
Received: from aserp3030.oracle.com (aserp3030.oracle.com [141.146.126.71])
        by aserp2120.oracle.com with ESMTP id 3091m13x9d-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Thu, 09 Apr 2020 21:30:47 +0000
Received: from pps.filterd (aserp3030.oracle.com [127.0.0.1])
        by aserp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 039LMolE003589;
        Thu, 9 Apr 2020 21:30:47 GMT
Received: from userv0122.oracle.com (userv0122.oracle.com [156.151.31.75])
        by aserp3030.oracle.com with ESMTP id 309ag5r05a-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Thu, 09 Apr 2020 21:30:47 +0000
Received: from abhmp0016.oracle.com (abhmp0016.oracle.com [141.146.116.22])
        by userv0122.oracle.com (8.14.4/8.14.4) with ESMTP id 039LUkQT024388;
        Thu, 9 Apr 2020 21:30:46 GMT
Received: from ban25x6uut29.us.oracle.com (/10.153.73.29)
        by default (Oracle Beehive Gateway v4.0)
        with ESMTP ; Thu, 09 Apr 2020 14:30:46 -0700
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com
Subject: [PATCH 1/3] KVM: nSVM: Check for CR0.CD and CR0.NW on VMRUN of nested
 guests
Date: Thu,  9 Apr 2020 16:50:33 -0400
Message-Id: <20200409205035.16830-2-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.20.1
In-Reply-To: <20200409205035.16830-1-krish.sadhukhan@oracle.com>
References: <20200409205035.16830-1-krish.sadhukhan@oracle.com>
MIME-Version: 1.0
X-Proofpoint-Virus-Version: vendor=nai engine=6000 definitions=9586
 signatures=668685
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 adultscore=1
 spamscore=0 malwarescore=0
 phishscore=0 mlxscore=0 bulkscore=0 mlxlogscore=959 suspectscore=1
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2003020000
 definitions=main-2004090151
X-Proofpoint-Virus-Version: vendor=nai engine=6000 definitions=9586
 signatures=668685
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 adultscore=0
 mlxlogscore=999 mlxscore=0
 priorityscore=1501 phishscore=0 suspectscore=1 bulkscore=0
 lowpriorityscore=0 impostorscore=0 malwarescore=0 clxscore=1015
 spamscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2003020000 definitions=main-2004090151
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

According to section "Canonicalization and Consistency Checks" in APM vol. 2,
the following guest state combination is illegal:

	"CR0.CD is zero and CR0.NW is set"

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 arch/x86/kvm/svm/nested.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/x86/kvm/svm/nested.c b/arch/x86/kvm/svm/nested.c
index 90a1ca9..38e4c2b 100644
--- a/arch/x86/kvm/svm/nested.c
+++ b/arch/x86/kvm/svm/nested.c
@@ -207,6 +207,10 @@ static bool nested_vmcb_checks(struct vmcb *vmcb)
 	if ((vmcb->save.efer & EFER_SVME) == 0)
 		return false;
 
+	if (((vmcb->save.cr0 & X86_CR0_CD) == 0) &&
+	    (vmcb->save.cr0 & X86_CR0_NW))
+		return false;
+
 	if ((vmcb->control.intercept & (1ULL << INTERCEPT_VMRUN)) == 0)
 		return false;
 

From patchwork Thu Apr  9 20:50:34 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 11482481
Return-Path: <SRS0=jZn3=5Z=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 589EE14B4
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 21:30:52 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 395F92083E
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 21:30:52 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=oracle.com header.i=@oracle.com
 header.b="pUM8stcJ"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727007AbgDIVav (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 9 Apr 2020 17:30:51 -0400
Received: from aserp2120.oracle.com ([141.146.126.78]:36658 "EHLO
        aserp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726917AbgDIVau (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 9 Apr 2020 17:30:50 -0400
Received: from pps.filterd (aserp2120.oracle.com [127.0.0.1])
        by aserp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 039LNvXD133470;
        Thu, 9 Apr 2020 21:30:48 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references : mime-version :
 content-transfer-encoding; s=corp-2020-01-29;
 bh=/6mnzP6MXdiweNB+aTC67Y8zflCbakbrTJoiQ62WKvE=;
 b=pUM8stcJKITevNSgsrgJqLlFEtJlu2x8M8xaMdZYEtWRaLpabHfC4yE/+QgG+HGY5EA1
 UWfT3Av2M1cbqbJGYNtrcvRRWitBoiViAGMS13AVPitmew2DIIyZLNRrxgFYcltijjUF
 br4m15XR2t4Sok/Xsw2j/P1eRJIQZVg4NRtGrJnJSxXwy87lIF+JtVYIkjc02/t9Jgy/
 WOCBe+jzmzOEorSKHFwBRBFy5+tVpVGs+y6GOJAM+RRkDiGyWz0Jx97EDrN8/lz44yrw
 6nRfz53UQuOrL+W7El5RhtDh+0WDjrUyAmgiWPxo8TlEC8vzx9La9Ect8C/4CvKfl9WD jw==
Received: from aserp3030.oracle.com (aserp3030.oracle.com [141.146.126.71])
        by aserp2120.oracle.com with ESMTP id 3091m13x9g-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Thu, 09 Apr 2020 21:30:48 +0000
Received: from pps.filterd (aserp3030.oracle.com [127.0.0.1])
        by aserp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 039LMnP1003484;
        Thu, 9 Apr 2020 21:30:47 GMT
Received: from userv0122.oracle.com (userv0122.oracle.com [156.151.31.75])
        by aserp3030.oracle.com with ESMTP id 309ag5r05k-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Thu, 09 Apr 2020 21:30:47 +0000
Received: from abhmp0016.oracle.com (abhmp0016.oracle.com [141.146.116.22])
        by userv0122.oracle.com (8.14.4/8.14.4) with ESMTP id 039LUkEG024391;
        Thu, 9 Apr 2020 21:30:46 GMT
Received: from ban25x6uut29.us.oracle.com (/10.153.73.29)
        by default (Oracle Beehive Gateway v4.0)
        with ESMTP ; Thu, 09 Apr 2020 14:30:46 -0700
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com
Subject: [PATCH 2/3] kvm-unit-tests: SVM: Add #defines for CR0.CD and CR0.NW
Date: Thu,  9 Apr 2020 16:50:34 -0400
Message-Id: <20200409205035.16830-3-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.20.1
In-Reply-To: <20200409205035.16830-1-krish.sadhukhan@oracle.com>
References: <20200409205035.16830-1-krish.sadhukhan@oracle.com>
MIME-Version: 1.0
X-Proofpoint-Virus-Version: vendor=nai engine=6000 definitions=9586
 signatures=668685
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 adultscore=0
 spamscore=0 malwarescore=0
 phishscore=0 mlxscore=0 bulkscore=0 mlxlogscore=894 suspectscore=15
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2003020000
 definitions=main-2004090151
X-Proofpoint-Virus-Version: vendor=nai engine=6000 definitions=9586
 signatures=668685
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 adultscore=0
 mlxlogscore=970 mlxscore=0
 priorityscore=1501 phishscore=0 suspectscore=15 bulkscore=0
 lowpriorityscore=0 impostorscore=0 malwarescore=0 clxscore=1015
 spamscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2003020000 definitions=main-2004090151
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 lib/x86/processor.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/x86/processor.h b/lib/x86/processor.h
index 3adddee..804673b 100644
--- a/lib/x86/processor.h
+++ b/lib/x86/processor.h
@@ -32,6 +32,8 @@
 #define X86_CR0_TS     0x00000008
 #define X86_CR0_WP     0x00010000
 #define X86_CR0_AM     0x00040000
+#define X86_CR0_NW     0x20000000
+#define X86_CR0_CD     0x40000000
 #define X86_CR0_PG     0x80000000
 #define X86_CR3_PCID_MASK 0x00000fff
 #define X86_CR4_TSD    0x00000004

From patchwork Thu Apr  9 20:50:35 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 11482485
Return-Path: <SRS0=jZn3=5Z=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 732ED913
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 21:32:53 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 5231E2083E
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 21:32:53 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=oracle.com header.i=@oracle.com
 header.b="enZFGIyB"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727074AbgDIVcv (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 9 Apr 2020 17:32:51 -0400
Received: from userp2120.oracle.com ([156.151.31.85]:41096 "EHLO
        userp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726666AbgDIVcv (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 9 Apr 2020 17:32:51 -0400
Received: from pps.filterd (userp2120.oracle.com [127.0.0.1])
        by userp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 039LN5ZN021407;
        Thu, 9 Apr 2020 21:32:49 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references : mime-version :
 content-transfer-encoding; s=corp-2020-01-29;
 bh=ihUhFhpj3fVrjo+Pa4wWIprtfvUQwzlDc9R70YsCDSM=;
 b=enZFGIyBbUJZDcVh0nAjG/LS345e9euUMCVEBFIeH5SbCusKlfwEMaZQMdGQV69wL60x
 FBEH3eWAEx0GATkEQ/RUFKNGYkRE7g5QPl7pTqdrjm8h68M6Kyl2glrTij2oozTg31Lp
 M4nwTKK6PFuBwUazAxfESvWcYaJiUydIHQR2chQacohTpZamKxvQyHDtWnxljzu6JDnK
 0wyJyWCenQ3HAtYj6sCEOQaDOnatJqmFb0dNUquVEEwpxg1qrcJK7I7IMJwB/JEf4Voz
 a3YAigmgehnosF4RAQFtHH4jn/vRCsLGXZ/LamSmIRKW6gIXhtMeh3KHVbKPdcdj0rUQ 8w==
Received: from aserp3020.oracle.com (aserp3020.oracle.com [141.146.126.70])
        by userp2120.oracle.com with ESMTP id 309gw4fthu-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Thu, 09 Apr 2020 21:32:48 +0000
Received: from pps.filterd (aserp3020.oracle.com [127.0.0.1])
        by aserp3020.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 039LLd5C124255;
        Thu, 9 Apr 2020 21:30:48 GMT
Received: from userv0122.oracle.com (userv0122.oracle.com [156.151.31.75])
        by aserp3020.oracle.com with ESMTP id 3091m9mk2p-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Thu, 09 Apr 2020 21:30:48 +0000
Received: from abhmp0016.oracle.com (abhmp0016.oracle.com [141.146.116.22])
        by userv0122.oracle.com (8.14.4/8.14.4) with ESMTP id 039LUkSs024394;
        Thu, 9 Apr 2020 21:30:47 GMT
Received: from ban25x6uut29.us.oracle.com (/10.153.73.29)
        by default (Oracle Beehive Gateway v4.0)
        with ESMTP ; Thu, 09 Apr 2020 14:30:46 -0700
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com
Subject: [PATCH 3/3] kvm-unit-tests: nSVM: Test CR0.CD and CR0.NW combination
 on VMRUN of nested guests
Date: Thu,  9 Apr 2020 16:50:35 -0400
Message-Id: <20200409205035.16830-4-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.20.1
In-Reply-To: <20200409205035.16830-1-krish.sadhukhan@oracle.com>
References: <20200409205035.16830-1-krish.sadhukhan@oracle.com>
MIME-Version: 1.0
X-Proofpoint-Virus-Version: vendor=nai engine=6000 definitions=9586
 signatures=668685
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 phishscore=0
 bulkscore=0 mlxscore=0
 malwarescore=0 spamscore=0 adultscore=0 suspectscore=15 mlxlogscore=927
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2003020000
 definitions=main-2004090151
X-Proofpoint-Virus-Version: vendor=nai engine=6000 definitions=9586
 signatures=668685
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0
 priorityscore=1501 bulkscore=0
 phishscore=0 lowpriorityscore=0 impostorscore=0 clxscore=1015
 suspectscore=15 malwarescore=0 spamscore=0 mlxlogscore=972 mlxscore=0
 adultscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2003020000 definitions=main-2004090151
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

According to section "Canonicalization and Consistency Checks" in APM vol. 2,
the following guest state combination is illegal:

        "CR0.CD is zero and CR0.NW is set"

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 x86/svm_tests.c | 28 +++++++++++++++++++++++++++-
 1 file changed, 27 insertions(+), 1 deletion(-)

diff --git a/x86/svm_tests.c b/x86/svm_tests.c
index 16b9dfd..8bdefc5 100644
--- a/x86/svm_tests.c
+++ b/x86/svm_tests.c
@@ -1352,15 +1352,41 @@ static void basic_guest_main(struct svm_test *test)
 
 static void svm_guest_state_test(void)
 {
+	test_set_guest(basic_guest_main);
+
+	/*
+	 * Un-setting EFER.SVME is illegal
+	 */
 	u64 efer_saved = vmcb->save.efer;
 	u64 efer = efer_saved;
 
-	test_set_guest(basic_guest_main);
 	report (svm_vmrun() == SVM_EXIT_VMMCALL, "EFER.SVME: %lx", efer);
 	efer &= ~EFER_SVME;
 	vmcb->save.efer = efer;
 	report (svm_vmrun() == SVM_EXIT_ERR, "EFER.SVME: %lx", efer);
 	vmcb->save.efer = efer_saved;
+
+	/*
+	 * Un-setting CR0.CD and setting CR0.NW is illegal combination
+	 */
+	u64 cr0_saved = vmcb->save.cr0;
+	u64 cr0 = cr0_saved;
+
+	cr0 |= X86_CR0_CD;
+	cr0 &= ~X86_CR0_NW;
+	vmcb->save.cr0 = cr0;
+	report (svm_vmrun() == SVM_EXIT_VMMCALL, "CR0: %lx", cr0);
+	cr0 |= X86_CR0_NW;
+	vmcb->save.cr0 = cr0;
+	report (svm_vmrun() == SVM_EXIT_VMMCALL, "CR0: %lx", cr0);
+	cr0 &= ~X86_CR0_NW;
+	cr0 &= ~X86_CR0_CD;
+	vmcb->save.cr0 = cr0;
+	report (svm_vmrun() == SVM_EXIT_VMMCALL, "CR0: %lx", cr0);
+	cr0 |= X86_CR0_NW;
+	vmcb->save.cr0 = cr0;
+	report (svm_vmrun() == SVM_EXIT_ERR, "CR0: %lx", cr0);
+	vmcb->save.cr0 = cr0_saved;
 }
 
 struct svm_test svm_tests[] = {
